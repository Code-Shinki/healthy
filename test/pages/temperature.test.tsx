import { cleanup, render, screen } from '@testing-library/react'
import { useRouter } from 'next/router'
import React from 'react'
import Temperature from '../../src/pages/temperature'
import AppMock from '../mocks/app-mock'
import { loggedInDemo, loggedInInit, nowLoading, unregistered } from '../mocks/globalStateMock'
import { nextRouterMock } from '../mocks/nextRouterMock'

// モック関数を作成
jest.mock('next/router', () => ({ useRouter: jest.fn() }))
jest.mock('react-apexcharts', () => {
  return {
    __esModule: true,
    // eslint-disable-next-line react/display-name
    default: () => {
      return <div />
    },
  }
})

describe(`体温グラフページ (/temperature)`, () => {
  afterEach(cleanup)

  test('スナップショット', async () => {
    // useRouterのメソッドをモック化する
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    const { asFragment } = render(
      <AppMock initializeState={loggedInDemo} drawerMenu={true}>
        <Temperature />
      </AppMock>
    )

    const tree = asFragment()
    expect(tree).toMatchSnapshot()

    // 再レンダーを待機
    expect(await screen.findByRole('heading', { level: 1 })).toBeInTheDocument()
  })

  test('ログイン状態を判別中', async () => {
    // ソースファイルのrouterに渡した引数をmockRouterで受け取る
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    render(
      <AppMock initializeState={nowLoading} drawerMenu={true}>
        <Temperature />
      </AppMock>
    )

    // 読み込み中にcomponents/atoms/spinner.tsxが表示されるか
    expect(screen.getByText('読み込んでいます')).toBeInTheDocument()
  })

  test('ログインしていないユーザー', async () => {
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    render(
      <AppMock initializeState={unregistered} drawerMenu={true}>
        <Temperature />
      </AppMock>
    )

    // next/routerでログインページへリダイレクトしているか
    expect(mockRouter.push).toHaveBeenCalledWith('/login')
  })

  test('初期データを持つログイン済みのユーザー', async () => {
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    render(
      <AppMock initializeState={loggedInInit} drawerMenu={true}>
        <Temperature />
      </AppMock>
    )

    // メニューが存在するか
    expect(screen.getByRole('button', { name: 'ホーム' })).toBeInTheDocument()

    // ページタイトルが存在するか
    expect(screen.getByRole('heading', { level: 1, name: /体温グラフ/i })).toBeInTheDocument()

    // データが規定数以上存在しないため注意が表示されるか
    expect(screen.getByText('データが不足しています')).toBeInTheDocument()
  })

  test('フルデータを持つログイン済みのユーザー', async () => {
    // useRouterのメソッドをモック化する
    const mockRouter = nextRouterMock
    ;(useRouter as jest.Mock).mockReturnValue(mockRouter)

    render(
      <AppMock initializeState={loggedInDemo} drawerMenu={true}>
        <Temperature />
      </AppMock>
    )

    // メニューが存在するか
    expect(screen.getByRole('button', { name: 'ホーム' })).toBeInTheDocument()

    // ページタイトルが存在するか
    expect(screen.getByRole('heading', { level: 1, name: /体温グラフ/i })).toBeInTheDocument()

    // グラフが存在するか (モック化しているのでラッパーを検出)
    expect(await screen.findByRole('region')).toHaveClass('fullGraph')
  })
})
