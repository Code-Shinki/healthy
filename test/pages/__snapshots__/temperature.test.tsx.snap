// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`体温グラフページ (/temperature) スナップショット 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-container-1"
  >
    <div
      class="makeStyles-root-2"
    >
      <h1
        class="title"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20h4V4h-4v16zm-6 0h4v-8H4v8zM16 9v11h4V9h-4z"
          />
        </svg>
        体温グラフ
      </h1>
      <div
        class="container"
      >
        <section
          aria-label="graph-wrapper"
          class="fullGraph"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiBottomNavigation-root root"
  >
    <button
      class="MuiButtonBase-root MuiBottomNavigationAction-root item"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiBottomNavigationAction-wrapper"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label"
        >
          ホーム
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiBottomNavigationAction-root item"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiBottomNavigationAction-wrapper"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3h-4.18C14.4 1.84 13.3 1 12 1c-1.3 0-2.4.84-2.82 2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-7 0c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1zm2 14H7v-2h7v2zm3-4H7v-2h10v2zm0-4H7V7h10v2z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label"
        >
          ヘルスデータ
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiBottomNavigationAction-root item"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiBottomNavigationAction-wrapper"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20h4V4h-4v16zm-6 0h4v-8H4v8zM16 9v11h4V9h-4z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label"
        >
          体温グラフ
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiBottomNavigationAction-root item"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiBottomNavigationAction-wrapper"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label"
        >
          ユーザー
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiBottomNavigationAction-root item"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiBottomNavigationAction-wrapper"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label"
        >
          ログアウト
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
